<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>IReNA_pipline • IReNA</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="IReNA_pipline">
<meta property="og:description" content="IReNA">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">IReNA</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/IReNA_pipline.html">IReNA_pipline</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="IReNA_pipline_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>IReNA_pipline</h1>
            
      
      
      <div class="hidden name"><code>IReNA_pipline.Rmd</code></div>

    </div>

    
    
<div id="irena" class="section level1">
<h1 class="hasAnchor">
<a href="#irena" class="anchor"></a>IReNA</h1>
<!-- badges: start -->
<!-- badges: end -->
<p>IReNA (Integrated Regulatory Network Analysis) is to reconstruct regulatory networks through integrating scRNA-seq and ATAC-seq data. Compared with other regulatory network analysis method (SCENIC), IReNA provides modularized regulatory network analysis to discover the biological significance of transcription factors and the regulatory role of each module throughout the process.</p>
<p>IReNA contains four main part to reconstruct regulatory network:</p>
<p>Part 1: Analyze scRNA-seq or bulk RNA-seq data to get basic regulatory relationships between transcription factors and genes.</p>
<p>Part 2: Use RcisTarget to refine regulatory relaionships</p>
<p>Part 3: Analyze ATAC-seq data to refine regulatory relationships</p>
<p>Part 4: Regultory network analysis.</p>
<p>If you <strong>have ATAC-seq data</strong>, use <strong>part 3</strong> to refine regulatory relationships. If you <strong>don’t have ATAC-seq data</strong>, use <strong>part 2</strong> to refine regulatory relaionships</p>
<p>If you use ATAC-seq to refine regulatory relationships, you also need a linux system PC or server with the following softwares: <a href="http://www.htslib.org/">samtools</a>, <a href="https://bedtools.readthedocs.io/en/latest/">bedtools</a> and <a href="https://meme-suite.org/meme/doc/fimo.html">fimo</a>.</p>
<p>This README contains: (1) Installation (2) test data download (3) Workflow of IReNA (4) ATAC-seq data preprocessing (5) IReNA input (6) Tutorial (7) citation (8) Help and Suggestion</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>1.Installation</h2>
<p>IReNA needs R version 4.0 or higher, <a href="http://bioconductor.org/">Bioconductor</a> version 3.12.</p>
<p>First install Bioconductor, open R and run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span>version <span class="op">=</span> <span class="st">"3.12"</span><span class="op">)</span></code></pre></div>
<p>Next, install a few Bioconductor dependencies that aren’t automatically installed:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'Rsamtools'</span>, <span class="st">'ChIPseeker'</span>, <span class="st">'monocle'</span>,
                       <span class="st">'RcisTarget'</span>, <span class="st">'RCy3'</span>, <span class="st">'clusterProfiler'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Then, install IReNA from GitHub:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span>
<span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"jiang-junyao/IReNA"</span><span class="op">)</span></code></pre></div>
<p>Finally, check whether IReNA was installed correctly, start a new R session and run:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">IReNA</span><span class="op">)</span>
<span class="co">#&gt; Loading required package: pheatmap</span>
<span class="co">#&gt; Loading required package: DDRTree</span>
<span class="co">#&gt; Loading required package: irlba</span>
<span class="co">#&gt; Loading required package: Matrix</span>
<span class="co">#&gt; Loading required package: RcisTarget</span>
<span class="co">#&gt; </span></code></pre></div>
</div>
<div id="test-data-download" class="section level2">
<h2 class="hasAnchor">
<a href="#test-data-download" class="anchor"></a>2.Test data download</h2>
<p>Test data used below can be download from <a href="https://github.com/jiang-junyao/IReNA-test-data" class="uri">https://github.com/jiang-junyao/IReNA-test-data</a>. If you want raw data of ATAC-seq to run <a href="https://github.com/jiang-junyao/ATAC-seq-pipline">ATAC-seq analysis pipline</a>, you can download it from <a href="https://www.ncbi.nlm.nih.gov/biosample?Db=biosample&amp;DbFrom=bioproject&amp;Cmd=Link&amp;LinkName=bioproject_biosample&amp;LinkReadableName=BioSample&amp;ordinalpos=1&amp;IdsFromResult=357084" class="uri">https://www.ncbi.nlm.nih.gov/biosample?Db=biosample&amp;DbFrom=bioproject&amp;Cmd=Link&amp;LinkName=bioproject_biosample&amp;LinkReadableName=BioSample&amp;ordinalpos=1&amp;IdsFromResult=357084</a></p>
</div>
<div id="atac-seq-data-preprocessing" class="section level2">
<h2 class="hasAnchor">
<a href="#atac-seq-data-preprocessing" class="anchor"></a>3.ATAC-seq data preprocessing</h2>
<p>If you use ATAC-seq data to refine regulatory relaionships (part3), you need to preprocess ATAC-seq raw data(fastq) to get bam, peak and footprints. So we provide <a href="https://github.com/jiang-junyao/ATAC-seq-pipline">ATAC-seq analysis pipline</a> for user to preprocess ATAC-seq data. The object of this pipline is to get bam file of each sample, peaks file of all samples and footprints of all samples as IReNA input. If you are familiar with ATAC-seq data analysis, you can do it as you like.</p>
</div>
<div id="irena-input" class="section level2">
<h2 class="hasAnchor">
<a href="#irena-input" class="anchor"></a>4.IReNA Input</h2>
<p>If you <strong>only use scRNA-seq or bulk RNA-seq</strong> data to run IReNA, you just need to input (3).Raw counts of scRNA-seq data, or Seurat object, or bulk RNA-seq expression profile and (6).Motif database</p>
<p>If you use <strong>both ATAC-seq data and scRNA-seq or bulk RNA-seq data</strong>, you need to input all the following files.</p>
<div id="bam-file-of-each-sample" class="section level3">
<h3 class="hasAnchor">
<a href="#bam-file-of-each-sample" class="anchor"></a>(1).Bam file of each sample</h3>
<p>Bam file of each sample can be generated by the step 11 in <a href="https://github.com/jiang-junyao/ATAC-seq-pipline">ATAC-seq analysis pipline</a>.</p>
</div>
<div id="peak-file-of-all-samples" class="section level3">
<h3 class="hasAnchor">
<a href="#peak-file-of-all-samples" class="anchor"></a>(2).Peak file of all samples</h3>
<p>Peaks file of all samples used here can be generated by the step 8 in <a href="https://github.com/jiang-junyao/ATAC-seq-pipline">ATAC-seq analysis pipline</a>.</p>
</div>
<div id="raw-counts-of-scrna-seq-data-or-seurat-object-or-bulk-rna-seq-expression-profile-" class="section level3">
<h3 class="hasAnchor">
<a href="#raw-counts-of-scrna-seq-data-or-seurat-object-or-bulk-rna-seq-expression-profile-" class="anchor"></a>(3).Raw counts of scRNA-seq data, or Seurat object, or bulk RNA-seq expression profile.</h3>
<p>IReNA provides function to load raw counts of scRNA-seq data, and return seurat object. If your data is <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/output/matrices">10X format</a>, set parameter datatype = 0. If your data is normal counts format(txt suffix), set parameter dayatype =1. If you have your own seurat object, you can skip load_counts, and use your own seurat object as ‘seurat_object’.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### load 10X counts</span>
<span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_counts.html">load_counts</a></span><span class="op">(</span><span class="st">'10X_data/sample1/'</span>, datatype <span class="op">=</span> <span class="fl">0</span><span class="op">)</span>
<span class="co">### load normal counts</span>
<span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_counts.html">load_counts</a></span><span class="op">(</span><span class="st">'test_data.txt'</span>,datatype <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
<span class="co">### read seurat object</span>
<span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="st">'seurat_object.rds'</span><span class="op">)</span></code></pre></div>
<p>If you use bulk RNA-seq data to get basic regulatory relationships, just input your bulk RNA-seq expression matrix as ‘expression_profile’, Use the same code as the scRNA-seq data to continue the analysis.</p>
</div>
<div id="footprints-file" class="section level3">
<h3 class="hasAnchor">
<a href="#footprints-file" class="anchor"></a>(4).Footprints file</h3>
<p>Footprints file can be generated by the step 10 in <a href="https://github.com/jiang-junyao/ATAC-seq-pipline">ATAC-seq analysis pipline</a>.</p>
</div>
<div id="reference-genome-of-your-species" class="section level3">
<h3 class="hasAnchor">
<a href="#reference-genome-of-your-species" class="anchor"></a>(5).Reference genome of your species</h3>
<p>Reference genome needs to be the same as that used for mapping, you can download it from <a href="https://hgdownload.soe.ucsc.edu/downloads.html">UCSC</a>.</p>
</div>
<div id="motif-database" class="section level3">
<h3 class="hasAnchor">
<a href="#motif-database" class="anchor"></a>(6).Motif database</h3>
<p>IReNA contains DNA motif datasets for four species (Homo sapiens, Mus musculus, Zebrafish and Chicken) derived from <a href="https://genexplain.com/transfac/">TRANSFAC 201803</a>. Following codes are used to call the motif dataset from TRANSFAC or user-defined motif dataset which should have the same format as these from TRANSFAC database.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">###call Mus musculus motif database</span>
<span class="va">motif1</span> <span class="op">&lt;-</span> <span class="va">Tranfac201803_Mm_MotifTFsF</span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">###call Homo sapiens motif database</span>
<span class="va">motif1</span> <span class="op">&lt;-</span> <span class="va">Tranfac201803_Hs_MotifTFsF</span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">###call Zebrafish motif database</span>
<span class="va">motif1</span> <span class="op">&lt;-</span> <span class="va">Tranfac201803_Zf_MotifTFsF</span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">###call Chicken motif database</span>
<span class="va">motif1</span> <span class="op">&lt;-</span> <span class="va">Tranfac201803_Ch_MotifTFsF</span></code></pre></div>
</div>
</div>
<div id="tutorial" class="section level2">
<h2 class="hasAnchor">
<a href="#tutorial" class="anchor"></a>5.Tutorial</h2>
<p>IReNA contains four main part to reconstruct regulatory network:</p>
<p>Part 1: Analyze scRNA-seq or bulk RNA-seq data to get basic regulatory relationships</p>
<p>Part 2: Use RcisTarget to refine regulatory relaionships</p>
<p>Part 3: Analyze ATAC-seq data to refine regulatory relationships</p>
<p>Part 4: Regultory network analysis.</p>
<p>If you <strong>have ATAC-seq data</strong>, use <strong>part3</strong> to refine regulatory relationships. If you <strong>don’t have ATAC-seq data</strong>, use <strong>part2</strong> to refine regulatory relaionships.</p>
<div id="part-1-analyze-scrna-seq-or-bulk-rna-seq-data-to-get-basic-regulatory-relationships" class="section level3">
<h3 class="hasAnchor">
<a href="#part-1-analyze-scrna-seq-or-bulk-rna-seq-data-to-get-basic-regulatory-relationships" class="anchor"></a>Part 1: Analyze scRNA-seq or bulk RNA-seq data to get basic regulatory relationships</h3>
<p>IReNA supports two input format: (i)path of raw counts, you can input path of raw counts and use function in GReNA to load data and convert it to Seurat object; (ii)Seurat object. After you upload the data, IReNA can calculate pseudotime according to R package monocle and add it to the metadata of Seurat object.</p>
<p>Our test seurat object only contains differentially expressed genes, so we set the parameter ‘DEG’ in add_pseudotime_DEG_filter() function as FALSE, if you seurat object contain all genes, please set this parameter as TRUE. In paraell, our test seurat object have been normalized, so we set the parameter ‘normlize1’ in add_pseudotime_DEG_filter() function as FALSE, if your seurat object only contains raw counts, please set this parameter as TRUE.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">###Read seurat_object</span>
<span class="fu"><a href="https://rdrr.io/r/base/load.html">load</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"test_seurat.rda"</span>, package <span class="op">=</span> <span class="st">"IReNA"</span><span class="op">)</span><span class="op">)</span>
<span class="va">seurat_object</span> <span class="op">&lt;-</span> <span class="va">test_seurat</span>
<span class="co">###calculate the pseudotime and return monocle object</span>
<span class="va">monocle_object</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_pseudotime.html">get_pseudotime</a></span><span class="op">(</span><span class="va">seurat_object</span><span class="op">)</span>
<span class="co">###Add pseudotime to the Seurat object</span>
<span class="va">seurat_with_time</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_pseudotime_DEG_filter.html">add_pseudotime_DEG_filter</a></span><span class="op">(</span><span class="va">seurat_object</span>, <span class="va">monocle_object</span>,DEG <span class="op">=</span> <span class="cn">FALSE</span>,normlize1 <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<p>Then, cells are divided into 50 bins across pseudotime. The bin is removed if all genes in this bin have no expression. Gene is filtered if fold change &lt; 0.01 (setting by the parameter FC). Then, genes will be clustered through K-means algorithm (K is setting by the parameter K1).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">###Get expression profiles ordered by pseudotime</span>
<span class="va">expression_profile</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_SmoothByBin_PseudotimeExp.html">get_SmoothByBin_PseudotimeExp</a></span><span class="op">(</span><span class="va">seurat_with_time</span>, Bin <span class="op">=</span> <span class="fl">50</span><span class="op">)</span>
<span class="co">###Filter noise and logFC in expression profile</span>
<span class="va">expression_profile_filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fileter_expression_profile.html">fileter_expression_profile</a></span><span class="op">(</span><span class="va">expression_profile</span>, FC<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span>
<span class="co">###K-means clustering</span>
<span class="va">clustering</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustering_Kmeans.html">clustering_Kmeans</a></span><span class="op">(</span><span class="va">expression_profile_filter</span>, K1<span class="op">=</span><span class="fl">4</span><span class="op">)</span>
<span class="co">#&gt; [1] "Perform k-means"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;  1  2  3  4 </span>
<span class="co">#&gt; 14 19 29 38 </span>
<span class="co">#&gt; [1] "Sort genes"</span>
<span class="co">#&gt; [1] "Revise outlier"</span>
<span class="co">#&gt; [1] "Number of outlier: 0" "Number of outlier: 0"</span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">clustering</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;         KmeansGroup FoldChangeQ95     SmExp1    SmExp5      SmExp6</span>
<span class="co">#&gt; FGFR3             1     0.4936206  1.2353418 2.0501186  0.30774396</span>
<span class="co">#&gt; IKBKAP            1     0.5857337 -0.6156551 1.0688363  0.32233389</span>
<span class="co">#&gt; SMARCE1           1     0.6312792  1.4589050 1.0649168 -0.00166819</span>
<span class="co">#&gt; CCNT2             1     0.6140669 -1.6394941 1.4556820  0.41294931</span>
<span class="co">#&gt; REST              1     0.7301245 -1.2891370 0.8269094 -0.12804089</span></code></pre></div>
<p>Visualize your clustering result through heatmap</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_kmeans_pheatmap.html">plot_kmeans_pheatmap</a></span><span class="op">(</span><span class="va">clustering</span>,ModuleColor1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">'#67C7C1'</span>,<span class="st">'#67C1E3'</span>,<span class="st">'#5BA6DA'</span>,<span class="st">'#EF9951'</span>,<span class="st">'#FFBF0F'</span>,<span class="st">'#C067A9'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>Adding Ensmble ID of the genes in the first column, then calculate the correlation of the gene pair and select gene pairs which contain at least one gene in transcription factors database and have absolute value of correlation larger than 0.6(setting by the parameter correlatio_filter).</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">###Add Ensembl ID as the first column of clustering results</span>
<span class="va">Kmeans_clustering_ENS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_ENSID.html">add_ENSID</a></span><span class="op">(</span><span class="va">clustering</span>, Spec1<span class="op">=</span><span class="st">'Hs'</span><span class="op">)</span>
<span class="va">Kmeans_clustering_ENS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;                  Symbol KmeansGroup FoldChangeQ95     SmExp1    SmExp5</span>
<span class="co">#&gt; ENSG00000068078   FGFR3           1     0.4936206  1.2353418 2.0501186</span>
<span class="co">#&gt; ENSG00000070061  IKBKAP           1     0.5857337 -0.6156551 1.0688363</span>
<span class="co">#&gt; ENSG00000073584 SMARCE1           1     0.6312792  1.4589050 1.0649168</span>
<span class="co">#&gt; ENSG00000082258   CCNT2           1     0.6140669 -1.6394941 1.4556820</span>
<span class="co">#&gt; ENSG00000084093    REST           1     0.7301245 -1.2891370 0.8269094</span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### Caculate the correlation</span>
<span class="va">motif1</span> <span class="op">&lt;-</span> <span class="va">Tranfac201803_Hs_MotifTFsF</span>
<span class="va">regulatory_relationships</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cor.html">get_cor</a></span><span class="op">(</span><span class="va">Kmeans_clustering_ENS</span>, motif <span class="op">=</span> <span class="va">motif1</span>, <span class="fl">0.6</span>, start_column <span class="op">=</span> <span class="fl">4</span><span class="op">)</span>
<span class="co">#&gt; [1] "get correlation for gene pairs"</span></code></pre></div>
</div>
<div id="part-2-use-rcistarget-to-refine-regulatory-relaionships-without-atac-seq-data" class="section level3">
<h3 class="hasAnchor">
<a href="#part-2-use-rcistarget-to-refine-regulatory-relaionships-without-atac-seq-data" class="anchor"></a>Part 2: Use RcisTarget to refine regulatory relaionships (without ATAC-seq data)</h3>
<p>For users who do not have ATAC-seq data, IReNA provides filter_regulation function (Based on RcisTarget) to refine regulation relaionships. Due to the limitations of RcisTarget, this function currently only supports three species (Hs, Mm and Fly). <strong>So if the species of your data is not included, and you don’t have ATAC-seq data, you can use unrefined regulatory relaionships to perform part4 analysis directly.</strong></p>
<p>Before run this function, you need to download Gene-motif rankings database from <a href="https://resources.aertslab.org/cistarget/" class="uri">https://resources.aertslab.org/cistarget/</a>, and set the Rankingspath1 as the path of downloaded Gene-motif rankings database. If you don’t know which database to choose, we suggest that using ‘hg19-500bp-upstream-7species.mc9nr’ for human, using ‘mm9-500bp-upstream-10species.mc9nr’ for mouse, using ‘dm6-5kb-upstream-full-tx-11species.mc8nr’ for Fly. You can download it manually, or use R code:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### Download Gene-motif rankings database</span>
<span class="va">featherURL</span> <span class="op">&lt;-</span> <span class="st">"https://resources.aertslab.org/cistarget/databases/homo_sapiens/hg19/refseq_r45/mc9nr/gene_based/hg19-tss-centered-10kb-7species.mc9nr.feather"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span><span class="op">(</span><span class="va">featherURL</span>, destfile<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html">basename</a></span><span class="op">(</span><span class="va">featherURL</span><span class="op">)</span><span class="op">)</span> <span class="co"># saved in current dir</span>
<span class="co">### Refine regulatory relaionships</span>
<span class="va">Rankingspath1</span> <span class="op">&lt;-</span> <span class="st">'hg19-500bp-upstream-7species.mc9nr1.feather'</span> <span class="co"># download from https://resources.aertslab.org/cistarget/</span>
<span class="va">filtered_regulatory_relationships</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/filter_regulation.html">filter_regulation</a></span><span class="op">(</span><span class="va">regulatory_relationships</span>, <span class="st">'Hs'</span>, <span class="va">Rankingspath1</span><span class="op">)</span></code></pre></div>
</div>
<div id="part-3-analyze-atac-seq-data-to-refine-regulatory-relationships-have-atac-seq-data" class="section level3">
<h3 class="hasAnchor">
<a href="#part-3-analyze-atac-seq-data-to-refine-regulatory-relationships-have-atac-seq-data" class="anchor"></a>Part 3: Analyze ATAC-seq data to refine regulatory relationships (have ATAC-seq data)</h3>
<p>For users who have ATAC-seq data, IReNA provides several functions to calculate related transcription factors of footprints with high FOS to refine regulatory relationships. We merge footprints whose distance is less than 4 and get sequence from each footprints based on reference genome through function get_merged_fasta(). Reference genome should be fasta/fa format, and you can download it from <a href="https://hgdownload.soe.ucsc.edu/downloads.html#alpaca" class="uri">https://hgdownload.soe.ucsc.edu/downloads.html#alpaca</a> or other genome database website.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">###merge footprints whose distance is less than 4</span>
<span class="va">filtered_footprints</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">'footprints.bed'</span>,sep <span class="op">=</span> <span class="st">'\t'</span><span class="op">)</span>
<span class="va">fastadir</span> <span class="op">&lt;-</span> <span class="st">'Genome/hg38.fa'</span> 
<span class="va">merged_fasta</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_merged_fasta.html">get_merged_fasta</a></span><span class="op">(</span><span class="va">filtered_footprints</span>,<span class="va">fastadir</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">merged_fasta</span>,<span class="st">'merged_footprints.fasta'</span>,row.names<span class="op">=</span><span class="cn">F</span>,quote<span class="op">=</span><span class="cn">F</span>,col.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></code></pre></div>
<p>In this step, because <a href="https://meme-suite.org/meme/doc/fimo.html">fimo</a> software only have linux version, and it takes too long to implement the corresponding function on Windows, we generate a shell script to run Fimo software in shell. If you are familiar with linux system and Fimo, you can write your own commands as you like.</p>
<p>First, you need to identify differentially expressed genes related motifs through motif_select() function, which can help you to reduce running time of the subsequent analysis process.</p>
<p>Before run find_motifs() function, you should set the following four parameters: (1) fimodir: path of fimo software, if you have added fimo to the environment variable, just set this argument as ‘fimo’. (2) outputdir1: output path of shell script. (3) outputdir: output path of fimo result. (4) motifdir: path of motif file, you can download it from <a href="https://genexplain.com/transfac/">TRANSFAC201803</a>. (5) sequencedir: path of sequence which generated by get_merged_fasta(). Please note that, at the end of outputdir and sequencedir must contain / symbol.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### Identify differentially expressed genes related motifs</span>
<span class="va">motif1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/motifs_select.html">motifs_select</a></span><span class="op">(</span><span class="va">Tranfac201803_Hs_MotifTFsF</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">Kmeans_clustering_ENS</span><span class="op">)</span><span class="op">)</span> <span class="co">###Kmeans_clustering_ENS was obtained in part1</span>
<span class="co">### run find_motifs()</span>
<span class="va">fimodir</span> <span class="op">&lt;-</span> <span class="st">'fimo'</span>
<span class="va">outputdir1</span> <span class="op">&lt;-</span> <span class="st">'D:/GIBH/IReNA2 R package/IReNA2/ATAC/fimo/'</span>
<span class="va">outputdir</span> <span class="op">&lt;-</span> <span class="st">'/public/home/user/fimo/output/'</span>
<span class="va">motifdir</span> <span class="op">&lt;-</span> <span class="st">'/public/home/user/fimo/Mememotif/'</span>
<span class="va">sequencedir</span> <span class="op">&lt;-</span> <span class="st">'/public/home/user/fimo/merged_footprints.fasta'</span>
<span class="fu"><a href="../reference/find_motifs.html">find_motifs</a></span><span class="op">(</span><span class="va">motif1</span>,step<span class="op">=</span><span class="fl">20</span>,<span class="va">fimodir</span>, <span class="va">outputdir1</span>, <span class="va">outputdir</span>, <span class="va">motifdir</span>, <span class="va">sequencedir</span><span class="op">)</span></code></pre></div>
<p>Then you need to use <a href="https://winscp.net/eng/download.php">winscp</a> or other related software to transfer script in outputdir1 to ‘/public/home/user/fimo’ directory in Linux system, and run following commands (If you make analysis in linux system, ignore transfer part).</p>
<pre><code>### run the following commands in the shell
cd /public/home/user/fimo/
mkdir output
sh ./fimo_all.sh</code></pre>
<p>After get the result of Fimo, you can use <a href="https://winscp.net/eng/download.php">winscp</a> or other related software to transfer all files in output directory (fimo result files) from linux to windos(If you use R in linux, please ignore this part). Then, we combine these Fimo consequence in Dir2. Notably Dir2 folder should only contain fimo result files. Next, we load the peaks file and overlap differential peaks and motif footprints through overlap_footprints_peaks() function</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">###Combine all footprints of motifs</span>
<span class="va">Dir2</span> <span class="op">&lt;-</span> <span class="st">'D:/GIBH/IReNA2 R package/IReNA2/ATAC/fimo/output/'</span>
<span class="va">combined</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/combine_footprints.html">combine_footprints</a></span><span class="op">(</span><span class="va">Dir2</span><span class="op">)</span>
<span class="va">peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.delim</a></span><span class="op">(</span><span class="st">'D:\\IReNA\\ATAC\\Peaks\\differential_peaks.bed'</span><span class="op">)</span>
<span class="va">overlapped</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/overlap_footprints_peaks.html">overlap_footprints_peaks</a></span><span class="op">(</span><span class="va">combined</span>,<span class="va">peaks</span><span class="op">)</span></code></pre></div>
<p>However, the running time of overlap_footprints() is too long, so it’s highly recommanded to use bedtools to do overlap in linux system. If you want to use bedtools to do overlap, you need to output ‘combined_footprints’ dataframe, and transfer it to shell.(If you make analysis in linux system, ignore transfer part)</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="co">### output combined_footprints</span></span>
<span id="cb21-2"><a href="#cb21-2"></a><span class="kw">write.table</span>(combined,<span class="st">'combined.txt'</span>,<span class="dt">quote =</span> F,<span class="dt">row.names =</span> F,<span class="dt">col.names =</span> F,<span class="dt">sep =</span> <span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb21-3"><a href="#cb21-3"></a><span class="co">### Transfer combied.txt and differential_peaks.bed to linux system, and then run the following commands in the shell</span></span>
<span id="cb21-4"><a href="#cb21-4"></a>bedtools intersect <span class="op">-</span>a combined.txt <span class="op">-</span>b differential_peaks.bed <span class="op">-</span>wa <span class="op">-</span>wb <span class="op">&gt;</span><span class="st"> </span>overlappd.txt</span></code></pre></div>
<p>Next, we intergrate bioconductor package <a href="https://bioconductor.org/packages/release/bioc/vignettes/ChIPseeker/inst/doc/ChIPseeker.html">ChIPseeker</a> to get footprint-related genes. Before we run get_related_genes(), we need to specify TxDb, which can be download from: <a href="http://bioconductor.org/packages/release/BiocViews.html#___TxDb"></a>. Kmeans_clustering_ENS used here was obtained in part1.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### If you make overlap by bedtools, read 'overlapped.txt' to R</span>
<span class="va">overlapped</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span><span class="op">(</span><span class="st">'overlapped.txt'</span><span class="op">)</span>
<span class="co">###get footprint-related genes</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span>
<span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span>
<span class="va">list1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_related_genes.html">get_related_genes</a></span><span class="op">(</span><span class="va">overlapped</span>,txdb <span class="op">=</span> <span class="va">txdb</span>,motif<span class="op">=</span><span class="va">Tranfac201803_Mm_MotifTFsF</span>,Species <span class="op">=</span> <span class="st">'Hs'</span><span class="op">)</span>
<span class="co">###Get candidate genes/TFs-related peaks</span>
<span class="va">list2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_related_peaks.html">get_related_peaks</a></span><span class="op">(</span><span class="va">list1</span>,<span class="va">Kmeans_clustering_ENS</span><span class="op">)</span>
<span class="co">### output filtered footprints</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">list2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,<span class="st">'filtered_footprints.bed'</span>, quote <span class="op">=</span> <span class="cn">F</span>, row.names <span class="op">=</span> <span class="cn">F</span>, col.names <span class="op">=</span> <span class="cn">F</span>, sep <span class="op">=</span> <span class="st">'\t'</span><span class="op">)</span></code></pre></div>
<p>Then, because of the size of original bam file is too large, so we need to use samtools to extract footprints realated regions in bam to reduce running time of function which analyze bam files in IReNA. (If you use our <a href="https://github.com/jiang-junyao/IReNA-test-data">test data</a>, just skip this step)</p>
<pre><code>### transfer filtered_footprints.bed to linux system and run the following codes
samtools view -hb -L filtered_footprint.bed SSC_patient1.bam &gt; SSC1_filter.bam
samtools view -hb -L filtered_footprint.bed SSC_patient2.bam &gt; SSC2_filter.bam
samtools view -hb -L filtered_footprint.bed esc.bam &gt; esc_filter.bam</code></pre>
<p>In this step, we count the cuts of each position in footrprints by wig_track(), and use these cuts to calculate the FOS of footprints to identify enriched TFs which determine the regulatory relationship. regulatory_relationships used here was calculated in part1.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">### calculate cuts of each each position in footprints</span>
<span class="va">bamfilepath1</span> <span class="op">&lt;-</span> <span class="st">'SSC1_filter.bam'</span>
<span class="va">bamfilepath2</span> <span class="op">&lt;-</span> <span class="st">'SSC2_filter.bam'</span>
<span class="va">bamfilepath3</span> <span class="op">&lt;-</span> <span class="st">'esc_filter.bam'</span>
<span class="va">cuts1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wig_track.html">wig_track</a></span><span class="op">(</span>bamfilepath <span class="op">=</span> <span class="va">bamfilepath1</span>,bedfile <span class="op">=</span> <span class="va">list2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="va">cuts2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wig_track.html">wig_track</a></span><span class="op">(</span>bamfilepath <span class="op">=</span> <span class="va">bamfilepath2</span>,bedfile <span class="op">=</span> <span class="va">list2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="va">cuts3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/wig_track.html">wig_track</a></span><span class="op">(</span>bamfilepath <span class="op">=</span> <span class="va">bamfilepath3</span>,bedfile <span class="op">=</span> <span class="va">list2</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="va">wig_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="va">cuts1</span>,<span class="va">cuts2</span>,<span class="va">cuts3</span><span class="op">)</span>
<span class="co">### get related genes of footprints with high FOS</span>
<span class="va">potential_regulation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Footprints_FOS.html">Footprints_FOS</a></span><span class="op">(</span><span class="va">wig_list</span>,<span class="va">list2</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>
<span class="co">### Use related genes of footprints with high FOS to refine regulatory relationships</span>
<span class="va">filtered_regulatory_relationships</span> <span class="op">&lt;-</span> <span class="va">regulatory_relationships</span><span class="op">[</span><span class="va">regulatory_relationships</span><span class="op">$</span><span class="va">TF</span> <span class="op">%in%</span> <span class="va">potential_regulation</span><span class="op">$</span><span class="va">TF</span> <span class="op">&amp;</span> <span class="va">regulatory_relationships</span><span class="op">$</span><span class="va">Target</span> <span class="op">%in%</span> <span class="va">potential_regulation</span><span class="op">$</span><span class="va">Target</span>,<span class="op">]</span></code></pre></div>
</div>
<div id="part-4-regulatory-network-analysis" class="section level3">
<h3 class="hasAnchor">
<a href="#part-4-regulatory-network-analysis" class="anchor"></a>Part 4: Regulatory network analysis</h3>
<p>After we get ‘filtered_regulatory_relationships’ and ‘Kmeans_clustering_ENS’, we can reconstruct regulatory network. Run network_analysis() to get regulatory, this step will generate a list which contain the following 9 dataframes:</p>
<p>(1)Cor_TFs.txt: list of expressed TFs in the gene networks.</p>
<p>(2)Cor_EnTFs.txt: list of TFs which significantly regulate gene modules (or enriched TFs).</p>
<p>(3)FOSF_RegMTF_Cor_EnTFs.txt: regulatory pairs in which the source gene is enriched TF.</p>
<p>(4)FOSF_RegMTF_Cor_EnTFs.txt: regulatory pairs in which both source gene and target gene are enriched TFs.</p>
<p>(5)FOSF_RegMTF_Cor_EnTFs.txt: regulatory pairs only including regulations within each module but not those between modules, in this step</p>
<p>(6)TF_list: enriched TFs which significantly regulate gene modules</p>
<p>(7)TF_module_regulation: details of enriched TFs which significantly regulate gene modules</p>
<p>(8)TF_network: regulatory network for enriched transcription factors of each module</p>
<p>(9)intramodular_network: intramodular regulatory network</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">TFs_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/network_analysis.html">network_analysis</a></span><span class="op">(</span><span class="va">regulatory_relationships</span>,<span class="va">Kmeans_cluster_Ens</span>,TFFDR1 <span class="op">=</span> <span class="fl">10</span>,TFFDR2 <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></code></pre></div>
<p>We can also make enrichment analysis for differentially expressed genes in each module. Before you run this function, you need to download the org.db for your species through BiocManager.</p>
<pre><code><a href="https://rdrr.io/r/base/library.html">### Download Homo sapiens org.db
#BiocManger::install('org.Hs.eg.db')
library(org.Hs.eg.db)
### Enrichment analysis
enrichment_KEGG &lt;- enrich_module(Kmeans_clustering_ENS, org.Hs.eg.db, 'KEGG')
#enrichment_GO &lt;- enrich_module(Kmeans_cluster_ENS, org.Hs.eg.db, 'GO')
head(enrichment_KEGG)</a></code></pre>
<p>You can visualize regulatory network for enriched transcription factors of each module through plot_network() function by setting type parameter as ‘TF’. This plot shows regulatory relationships between transcription factors in different modules that significantly regulate other modules. The size of each vertex determine the significance of this transcription factor. Yellow edges are positive regulation,grey edges are negative regulation.</p>
<pre><code><a href="../reference/plot_tf_network.html">plot_tf_network(TFs_list)</a></code></pre>
<p>You can visualize intramodular network with enriched function through plot_intramodular_network() function by setting enrichment parameter as consequence of enrich_module(). This plot shows enriched transcription factors(has the most edges), enriched fcuntion(has the highest -log10qvalue) and regulatory relationships between each modules.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_intramodular_network.html">plot_intramodular_network</a></span><span class="op">(</span><span class="va">list1</span>,<span class="va">enrichment_KEGG</span>,layout <span class="op">=</span> <span class="st">'circle'</span><span class="op">)</span></code></pre></div>
<p>It is strongly recommended to use Cytoscape(downloading link: <a href="https://cytoscape.org/download.html" class="uri">https://cytoscape.org/download.html</a>) to display the regulatory networks. We provide a function that can provide different Cytoscape styles. You need to intall and open Cytoscape before running the function.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">###optional: display the network in cytoscape, open cytoscape before running this function</span>
<span class="fu"><a href="../reference/initiate_cy.html">initiate_cy</a></span><span class="op">(</span><span class="va">TFs_list</span>, layout1<span class="op">=</span><span class="st">'degree-circle'</span>, type<span class="op">=</span><span class="st">'TF'</span><span class="op">)</span>
<span class="fu"><a href="../reference/initiate_cy.html">initiate_cy</a></span><span class="op">(</span><span class="va">TFs_list</span>, layout1<span class="op">=</span><span class="st">'grid'</span>, type<span class="op">=</span><span class="st">'module'</span><span class="op">)</span></code></pre></div>
<p>This is the picture we processed through cytoscape, which can show the regulatory relationship of modularized transcription factors.</p>
</div>
</div>
<div id="citation" class="section level2">
<h2 class="hasAnchor">
<a href="#citation" class="anchor"></a>6.Citation</h2>
<p>If you use IReNA package, please cite the following Science paper: <a href="https://science.sciencemag.org/content/370/6519/eabb8598" class="uri">https://science.sciencemag.org/content/370/6519/eabb8598</a>.</p>
</div>
<div id="help-and-suggestion" class="section level2">
<h2 class="hasAnchor">
<a href="#help-and-suggestion" class="anchor"></a>7.Help and Suggestion</h2>
<p>If you have any question, comment or suggestion, please use github issue tracker to report coding related issues of CellChat or contact <a href="mailto:jiangjunyao789@163.com" class="email">jiangjunyao789@163.com</a>. I will answer you timely, and please remind me again if you have not received response more than three days.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Jiang Jun yao.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
